File: euler31/change.go
package main

import "fmt"

func main() {
	total := 5000
	coins := []int{200, 100, 50, 20, 10, 5, 2, 1}
	ways := make_change(total, coins)
	fmt.Println(ways)
}

// find ways to make change for the target_amount using coins of the given
// denominations
func make_change(target_amount int, coins []int) int {

	// this func merely wraps and supports (with memoization) the central
	// recursive alg

	type args struct {
		total_so_far, coin_index int
	}

	memo := map[args]int{}

	var recur func(total_so_far, coin_index int) int
	recur = func(total_so_far, coin_index int) int {

		// memoization shortcut
		if v, ok := memo[args{total_so_far, coin_index}]; ok {
			return v
		}

		// the alg proper
		coin_val := coins[coin_index]
		if total_so_far == target_amount || coin_val == 1 {
			return 1
		}
		will_fit := (target_amount - total_so_far) / coin_val
		sum := 0
		for n := 0; n < will_fit+1; n++ {
			sum += recur(total_so_far+coin_val*n, coin_index+1)
		}

		// save result and return
		memo[args{total_so_far, coin_index}] = sum
		return sum
	}
	return recur(0, 0)
}

File: euler31/change.py
US_currency = [100, 50, 25, 10, 5, 1]
US_target_amount = 100

english_currency = [200, 100, 50, 20, 10, 5, 2, 1]
english_target_amount = 200

coin = english_currency
target_amount = english_target_amount


def recur(total_so_far, coin_index):
    coin_value = coin[coin_index]
    if total_so_far == target_amount:
        return 1
    if coin_value == 1:
        return 1
    will_fit = (target_amount - total_so_far) / coin_value
    return sum(recur(total_so_far + coin_value * n, coin_index + 1)
               for n in range(will_fit + 1))


print recur(0, 0)
